# Esercizio1

Si consideri il dataset delle piogge mensili (mm) nei sette anni dal 2009 al 2015, che ha il seguente formato: 13 record di campi separati da tabulazione di cui il primo è il record di intestazione degli anni (composto da 7 campi) e gli altri 12 sono i record delle piogge mensili (un record per ciascun mese).
Ogni record è composto di 8 campi, di cui il primo è il nome del mese e i rimanenti 7 sono le piogge mensili lungo gli anni.

Si richiede di determinare:

- per ogni mese, la pioggia media lungo gli anni
- per ogni anno, la pioggia totale
- per ogni anno, il numero di mesi con pioggia oltre una certa soglia (parametro in input)

---

Parametri di input:
- dataset delle piogge
- soglia minima di pioggia

---

### Parametri di input

In [1]:
input_dataset = './input-piogge.txt'
rain_threshold = 100

### Definizione di una funzione che conta i valori di una lista oltre una certa soglia

In [4]:
def count_values_greater_than(list_of_values, threshold):
    return [value > threshold  for value in list_of_values].count(True)

### Lettura del dataset

- Funzione `open(file_name, 'r')` per ottenere l'*handle* in lettura al file
- Metodo `readlines()` per leggere il canale invocante in una lista di stringhe
- Metodo `close()` per chiudere l'*handle* invocante

In [8]:
with open(input_dataset, 'r') as input_file:
    file_rows = input_file.readlines()

In [9]:
file_rows

['\t\t\t2009\t2010\t2011\t2012\t2013\t2014\t2015\n',
 'Gennaio\t\t75\t\t63\t\t65\t\t50\t\t77\t\t66\t\t69\n',
 'Febbraio\t64\t\t65\t\t65\t\t67\t\t50\t\t54\t\t58\n',
 'Marzo\t\t81\t\t77\t\t73\t\t80\t\t83\t\t89\t\t100\n',
 'Aprile\t\t89\t\t90\t\t85\t\t90\t\t90\t\t84\t\t90\n',
 'Maggio\t\t120\t\t129\t\t113\t\t120\t\t135\t\t117\t\t130\n',
 'Giugno\t\t113\t\t99\t\t116\t\t114\t\t111\t\t119\t\t100\n',
 'Luglio\t\t111\t\t105\t\t98\t\t112\t\t113\t\t102\t\t100\n',
 'Agosto\t\t129\t\t131\t\t120\t\t111\t\t141\t\t130\t\t126\n',
 'Settembre\t90\t\t85\t\t101\t\t88\t\t89\t\t94\t\t91\n',
 'Ottobre\t\t109\t\t122\t\t103\t\t119\t\t98\t\t101\t\t107\n',
 'Novembre\t111\t\t121\t\t101\t\t104\t\t121\t\t115\t\t104\n',
 'Dicembre\t56\t\t67\t\t44\t\t58\t\t61\t\t64\t\t58']

### Costruzione della lista degli anni rilevati

In [10]:
header = file_rows.pop(0)

In [11]:
header

'\t\t\t2009\t2010\t2011\t2012\t2013\t2014\t2015\n'

In [12]:
file_rows

['Gennaio\t\t75\t\t63\t\t65\t\t50\t\t77\t\t66\t\t69\n',
 'Febbraio\t64\t\t65\t\t65\t\t67\t\t50\t\t54\t\t58\n',
 'Marzo\t\t81\t\t77\t\t73\t\t80\t\t83\t\t89\t\t100\n',
 'Aprile\t\t89\t\t90\t\t85\t\t90\t\t90\t\t84\t\t90\n',
 'Maggio\t\t120\t\t129\t\t113\t\t120\t\t135\t\t117\t\t130\n',
 'Giugno\t\t113\t\t99\t\t116\t\t114\t\t111\t\t119\t\t100\n',
 'Luglio\t\t111\t\t105\t\t98\t\t112\t\t113\t\t102\t\t100\n',
 'Agosto\t\t129\t\t131\t\t120\t\t111\t\t141\t\t130\t\t126\n',
 'Settembre\t90\t\t85\t\t101\t\t88\t\t89\t\t94\t\t91\n',
 'Ottobre\t\t109\t\t122\t\t103\t\t119\t\t98\t\t101\t\t107\n',
 'Novembre\t111\t\t121\t\t101\t\t104\t\t121\t\t115\t\t104\n',
 'Dicembre\t56\t\t67\t\t44\t\t58\t\t61\t\t64\t\t58']

In [21]:
years = header.split()

In [22]:
years

['2009', '2010', '2011', '2012', '2013', '2014', '2015']

### Costruzione della lista dei mesi

In [31]:
months = [row.split()[0] for row in file_rows]

In [32]:
months

['Gennaio',
 'Febbraio',
 'Marzo',
 'Aprile',
 'Maggio',
 'Giugno',
 'Luglio',
 'Agosto',
 'Settembre',
 'Ottobre',
 'Novembre',
 'Dicembre']

### Costruzione della matrice delle piogge

In [38]:
rains_per_month = [list(map(int, row.split()[1:])) for row in file_rows]

In [39]:
rains_per_month

[[75, 63, 65, 50, 77, 66, 69],
 [64, 65, 65, 67, 50, 54, 58],
 [81, 77, 73, 80, 83, 89, 100],
 [89, 90, 85, 90, 90, 84, 90],
 [120, 129, 113, 120, 135, 117, 130],
 [113, 99, 116, 114, 111, 119, 100],
 [111, 105, 98, 112, 113, 102, 100],
 [129, 131, 120, 111, 141, 130, 126],
 [90, 85, 101, 88, 89, 94, 91],
 [109, 122, 103, 119, 98, 101, 107],
 [111, 121, 101, 104, 121, 115, 104],
 [56, 67, 44, 58, 61, 64, 58]]

In [40]:
import numpy as np

In [43]:
#help(np)

In [45]:
rains_per_month = np.array(rains_per_month)

In [46]:
rains_per_month

array([[ 75,  63,  65,  50,  77,  66,  69],
       [ 64,  65,  65,  67,  50,  54,  58],
       [ 81,  77,  73,  80,  83,  89, 100],
       [ 89,  90,  85,  90,  90,  84,  90],
       [120, 129, 113, 120, 135, 117, 130],
       [113,  99, 116, 114, 111, 119, 100],
       [111, 105,  98, 112, 113, 102, 100],
       [129, 131, 120, 111, 141, 130, 126],
       [ 90,  85, 101,  88,  89,  94,  91],
       [109, 122, 103, 119,  98, 101, 107],
       [111, 121, 101, 104, 121, 115, 104],
       [ 56,  67,  44,  58,  61,  64,  58]])

### Trasposta della matrice delle piogge

In [48]:
rains_per_year = np.transpose(rains_per_month)

In [49]:
rains_per_year

array([[ 75,  64,  81,  89, 120, 113, 111, 129,  90, 109, 111,  56],
       [ 63,  65,  77,  90, 129,  99, 105, 131,  85, 122, 121,  67],
       [ 65,  65,  73,  85, 113, 116,  98, 120, 101, 103, 101,  44],
       [ 50,  67,  80,  90, 120, 114, 112, 111,  88, 119, 104,  58],
       [ 77,  50,  83,  90, 135, 111, 113, 141,  89,  98, 121,  61],
       [ 66,  54,  89,  84, 117, 119, 102, 130,  94, 101, 115,  64],
       [ 69,  58, 100,  90, 130, 100, 100, 126,  91, 107, 104,  58]])

### Piogge medie mensili

#### a) Costruzione della lista delle piogge medie mensili

In [51]:
avg_rain_list = [np.mean(row) for row in rains_per_month]

In [52]:
avg_rain_list

[66.42857142857143,
 60.42857142857143,
 83.28571428571429,
 88.28571428571429,
 123.42857142857143,
 110.28571428571429,
 105.85714285714286,
 126.85714285714286,
 91.14285714285714,
 108.42857142857143,
 111.0,
 58.285714285714285]

#### b) Output delle piogge medie mensili

Produrre il seguente output (stdout)

    GEN -> 66.42857142857143
    FEB -> 60.42857142857143
    MAR -> 83.28571428571429
    APR -> 88.28571428571429
    MAG -> 123.42857142857143
    GIU -> 110.28571428571429
    LUG -> 105.85714285714286
    AGO -> 126.85714285714286
    SET -> 91.14285714285714
    OTT -> 108.42857142857143
    NOV -> 111.0
    DIC -> 58.285714285714285

In [54]:
months = [month.upper()[:3] for month in months]

In [55]:
months

['GEN',
 'FEB',
 'MAR',
 'APR',
 'MAG',
 'GIU',
 'LUG',
 'AGO',
 'SET',
 'OTT',
 'NOV',
 'DIC']

In [58]:
avg_output_list = list(zip(months, avg_rain_list))

In [60]:
for (month, avg_rain) in avg_output_list:
    print(month, ' -> ', avg_rain)

GEN  ->  66.42857142857143
FEB  ->  60.42857142857143
MAR  ->  83.28571428571429
APR  ->  88.28571428571429
MAG  ->  123.42857142857143
GIU  ->  110.28571428571429
LUG  ->  105.85714285714286
AGO  ->  126.85714285714286
SET  ->  91.14285714285714
OTT  ->  108.42857142857143
NOV  ->  111.0
DIC  ->  58.285714285714285


### Piogge totali annue

#### a) Costruzione della lista delle piogge totali annue

In [63]:
year_total_rain = [sum(row) for row in rains_per_year]

In [64]:
year_total_rain

[1148, 1154, 1084, 1113, 1169, 1135, 1133]

---

#### b) Ouput delle piogge totali annue

Produrre il seguente output (stdout)

    2009 -> 1148
    2010 -> 1154
    2011 -> 1084
    2012 -> 1113
    2013 -> 1169
    2014 -> 1135
    2015 -> 1133

In [67]:
total_ouput_list = list(zip(years, year_total_rain))

In [68]:
for (year, total_rain) in total_ouput_list:
    print(year, ' -> ', total_rain)

2009  ->  1148
2010  ->  1154
2011  ->  1084
2012  ->  1113
2013  ->  1169
2014  ->  1135
2015  ->  1133


### Numero di mesi per anno con pioggia oltre la soglia minima

#### a) Costruzione della lista del numero di mesi (per anno) con pioggia oltre la soglia minima

In [72]:
year_total_rain = [count_values_greater_than(row, rain_threshold) for row in rains_per_year]

In [73]:
year_total_rain

[6, 5, 6, 6, 5, 6, 4]

#### b) Output del numero di mesi (per anno) con pioggia oltre la soglia minima

Produrre il seguente output (stdout)

    2009 -> 6
    2010 -> 5
    2011 -> 6
    2012 -> 6
    2013 -> 5
    2014 -> 6
    2015 -> 4

In [75]:
num_ouput_list = list(zip(years, year_total_rain))

In [76]:
for (year, num_rain) in num_ouput_list:
    print(year, ' -> ', num_rain)

2009  ->  6
2010  ->  5
2011  ->  6
2012  ->  6
2013  ->  5
2014  ->  6
2015  ->  4
